go/rpcclientlib/obscuro_client.go

```diff
@@ -30,28 +47,132 @@ type Client interface {
 }
 
 func NewClient(address string) Client {
+	// todo: this is a convenience for testnet but needs to replaced by a parameter and/or retrieved from the target host
+	enclPubECDSA, err := crypto.DecompressPubkey(common.Hex2Bytes(enclavePublicKeyHex))
+	if err != nil {
+		panic(err)
+	}
+	enclavePublicKey := ecies.ImportECDSAPublic(enclPubECDSA)
+
 	rpcClient, err := rpc.Dial(http + address)
 	if err != nil {
 		log.Panic("could not create RPC client on %s. Cause: %s", http+address, err)
 	}
 
 	return &clientImpl{
-		rpcClient: rpcClient,
+		rpcClient:        rpcClient,
+		enclavePublicKey: enclavePublicKey,
 	}
 }
 
+// Call handles JSON rpc requests - if the method is sensitive it will encrypt the args before sending the request and
+//		then decrypts the response before returning
+// The result must be a pointer so that package json can unmarshal into it. You
+// 	can also pass nil, in which case the result is ignored.
 func (c *clientImpl) Call(result interface{}, method string, args ...interface{}) error {
-	return c.rpcClient.Call(&result, method, args...)
+	if !isSensitive(method) {
+		// for non-sensitive methods we just delegate directly to the geth RPC client
+		return c.rpcClient.Call(&result, method, args...)
+	}
+
+	var encryptedParams interface{}
+	var err error
+	encryptedParams, err = c.encryptArgs(args)
+	if err != nil {
+		return fmt.Errorf("failed to encrypt args for rpc call - %w", err)
+	}
+
+	var rawResult []byte
+	if result == nil {
+		// we set result receiver to nil if caller set nil since no return needs to be set
+		rawResult = nil
+	}
+	err = c.rpcClient.Call(&rawResult, method, encryptedParams)
+	if err != nil {
+		return fmt.Errorf("rpc call failed - %w", err)
+	}
+
+	if result == nil {
+		// caller not interested in result, we're done
+		return nil
+	}
+
+	var resultMap map[string]interface{}
+	err = json.Unmarshal(rawResult, &resultMap)
+	if err != nil {
+		return err
+	}
+
+	encResult, f := resultMap[RespJSONKeyResult]
+	if !f {
+		return nil
+	}
+
+	// method is sensitive, so we decrypt it before unmarshalling the result
+	decrypted, err := c.decryptResponse(encResult)
+	if err != nil {
+		return fmt.Errorf("failed to decrypt args for rpc call - %w", err)
+	}
+
+	err = json.Unmarshal(decrypted, result)
+	if err != nil {
+		return fmt.Errorf("failed to unmarshal rpc response to json: %w", err)
+	}
+
+	return nil
+}
+
+func (c *clientImpl) encryptArgs(args interface{}) ([]byte, error) {
+	if args == nil {
+		return nil, nil
+	}
+	paramsJSON, err := json.Marshal(args)
+	if err != nil {
+		return nil, fmt.Errorf("could not json encode request params: %w", err)
+	}
+	encryptedParams, err := ecies.Encrypt(rand.Reader, c.enclavePublicKey, paramsJSON, nil, nil)
+	if err != nil {
+		return nil, fmt.Errorf("could not encrypt request params with enclave public key: %w", err)
+	}
+	return encryptedParams, nil
+}
+
+func (c *clientImpl) decryptResponse(resultBlob interface{}) ([]byte, error) {
```

Comment:
Check out the changes I had to make to the decryption on `main`. Some responses are just strings, but some are JSON lists. We ran into issues before because the transaction receipts were being returned as double-quoted json strings.